File: Earthshake/EarthshakeItem.m
//
//  EarthshakeItem.m
//  Earthshake
//
//  Created by Andrey Paunov on 2016-03-24.
//  Copyright © 2016 Andrey Paunov. All rights reserved.
//

#import "EarthshakeItem.h"

#define kPlace          @"place"
#define kMagnitude      @"mag"
#define kTime           @"time"
#define kTimeZone       @"tz"
#define kFeatureType    @"type"
#define kUrl            @"url"
#define kTsunami        @"tsunami"
#define kCoordinates    @"coordinates"

@implementation EarthshakeItem

// TODO: REFACTOR THE FORMATTERS TO IMPROVE PERFORMENCE

- (NSString *)place
{
    NSString *place = [self.properties objectForKey:kPlace];
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"^(\\d*\\s*(km)\\s*[NSEW]*\\s*(of)\\s*)" options:0 error:nil];
    NSTextCheckingResult *match = [regex firstMatchInString:place options:0 range:NSMakeRange(0, place.length)];
    NSRange range = [match rangeAtIndex:1];
    NSString *placeTrim = place;

    if (range.length)
    {
        placeTrim = [place substringFromIndex:range.location + range.length];
    }

    return placeTrim;
}

- (NSNumber *)magnitude
{
    NSNumber *magnitude = [NSNumber numberWithDouble:[[self.properties objectForKey:kMagnitude] doubleValue]];
    return magnitude;
}

- (NSString *)date
{
    double totalSeconds = [[self.properties objectForKey:kTime] doubleValue];

    NSDate *date = [NSDate dateWithTimeIntervalSince1970:(totalSeconds / 1000)];

    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy'/'MM'/'dd"];

    NSString *formattedDate = [dateFormatter stringFromDate:date];

    return formattedDate;
}

- (NSString *)time
{
    double totalSeconds = [[self.properties objectForKey:kTime] doubleValue];

    NSDate *date = [NSDate dateWithTimeIntervalSince1970:(totalSeconds / 1000)];

    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"HH':'mm z"];

    NSString *formattedDate = [dateFormatter stringFromDate:date];

    return formattedDate;
}

- (BOOL)isEarthquake
{
    BOOL earthquake = NO;
    NSString *type = [self.properties objectForKey:kFeatureType];

    if ([type isEqualToString:@"earthquake"])
    {
        earthquake = YES;
    }

    return earthquake;
}

- (BOOL)isQuarry
{
    BOOL quarry = NO;
    NSString *type = [self.properties objectForKey:kFeatureType];
    
    if ([type isEqualToString:@"quarry"])
    {
        quarry = YES;
    }
    
    return quarry;
}

- (NSString *)detailURLString
{
    NSString *urlString = [self.properties objectForKey:kUrl];
    return urlString;
}

- (CLLocationCoordinate2D)epicenter
{
    NSArray *coordinates = [self.geometry objectForKey:kCoordinates];

    CLLocationCoordinate2D epicenter = CLLocationCoordinate2DMake([[coordinates objectAtIndex:1] doubleValue], [[coordinates objectAtIndex:0] doubleValue]);

    return epicenter;
}

- (NSNumber *)tsunami
{
    NSNumber *tsunami = [self.properties objectForKey:kTsunami];
    return tsunami;
}

@end

File: Earthshake/EarthshakeItem.swift
//
//  EarthshakeItem.swift
//  Earthshake
//
//  Created by Andrey on 2016-12-28.
//  Copyright © 2016 AP. All rights reserved.
//

import Foundation
import MapKit

class EarthshakeItem : NSObject
{
    let kPlace = "place"
    let kMagnitude = "mag"
    let kTime = "time"
    let kTimeZone = "tz"
    let kFeatureType = "type"
    let kUrl = "url"
    let kTsunami = "tsunami"
    let kCoordinates = "coordinates"

    var properties = [AnyHashable: Any]()
    var geometry = [AnyHashable: Any]()
    var featureType = ""
    var earthshakeId = ""
    var title = ""
    var place: String = ""
    var magnitude: NSNumber!
    var date = ""
    var time = ""
    var detailURLString = ""
    var epicenter = CLLocationCoordinate2D()
    var tsunami: Int!

    func populateData(_ properties: [AnyHashable: Any], _ geometry: [AnyHashable: Any])
    {
        self.properties = properties
        self.geometry = geometry

        place = extractPlace()
        magnitude = extractMagnitude()
        date = extractDate()
        time = extractTime()
        detailURLString = extractDetailURLString()
        epicenter = extractEpicenter()
        tsunami = extractTsunami()
    }
    
    func extractPlace() -> String
    {
        var place = (properties[kPlace] as! String)
        let regex = try! NSRegularExpression(pattern: "^(\\d*\\s*(km)\\s*[NSEW]*\\s*(of)\\s*)", options: [])
        let match = regex.firstMatch(in: place, options: [], range: NSRange(location: 0, length: place.characters.count))
        let range = match?.rangeAt(1)
        var placeTrim = place

        if (range?.length) != nil && (range?.length)! > 0
        {
            placeTrim = place.substring(from: place.index(place.startIndex, offsetBy: (range?.location)! + (range?.length)!))
        }
        else
        {
            placeTrim = place
        }

        return placeTrim
    }

    func extractMagnitude() -> NSNumber
    {
        let magnitude = (properties[kMagnitude] as! NSNumber)
        return magnitude
    }

    func extractDate() -> String
    {
        let totalSeconds: Double = (properties[kTime] as! Double)
        let date = Date(timeIntervalSince1970: (totalSeconds / 1000))
        let dateFormatter = DateFormatter()

        dateFormatter.dateFormat = "yyyy'/'MM'/'dd"

        let formattedDate = dateFormatter.string(from: date)
        return formattedDate
    }

    func extractTime() -> String
    {
        let totalSeconds: Double = (properties[kTime] as! Double)
        let date = Date(timeIntervalSince1970: (totalSeconds / 1000))
        let dateFormatter = DateFormatter()
        
        dateFormatter.dateFormat = "HH':'mm z"
        
        let formattedDate = dateFormatter.string(from: date)
        return formattedDate
    }

    func extractDetailURLString() -> String
    {
        let urlString = (properties[kUrl] as! String)
        return urlString
    }

    func extractEpicenter() -> CLLocationCoordinate2D
    {
        var coordinates = (geometry[kCoordinates] as! [Double])
        let epicenter = CLLocationCoordinate2DMake(coordinates[1], coordinates[0])
        return epicenter
    }

    func extractTsunami() -> Int
    {
        let tsunami = (properties[kTsunami] as! Int)
        return tsunami
    }
    
    func isEarthquake() -> Bool
    {
        var earthquake = false
        let type = (properties[kFeatureType] as! String)
        
        if (type == "earthquake")
        {
            earthquake = true
        }
        
        return earthquake
    }
    
    func isQuarry() -> Bool
    {
        var quarry = false
        let type = (properties[kFeatureType] as! String)
        
        if (type == "quarry")
        {
            quarry = true
        }
        
        return quarry
    }
}
